File: hdf-mappings/test/cci.test.ts
import { describe, it, expect, vi } from 'vitest';
import {
  getCCIDescription,
  getCCINistMappings,
  getAllCCIIds,
  cciExists,
  getNistCCIMappings,
  nistToCci,
} from '../src/cci/index.js';

describe('CCI Mapping Functions', () => {
  describe('getCCIDescription', () => {
    it('should return description for valid CCI ID', () => {
      const desc = getCCIDescription('CCI-000001');
      expect(desc).toBeDefined();
      expect(desc).toContain('access control policy');
    });

    it('should return undefined for invalid CCI ID', () => {
      const desc = getCCIDescription('CCI-999999');
      expect(desc).toBeUndefined();
    });

    it('should handle empty string', () => {
      const desc = getCCIDescription('');
      expect(desc).toBeUndefined();
    });

    it('should be case-sensitive', () => {
      const desc = getCCIDescription('cci-000001');
      expect(desc).toBeUndefined();
    });
  });

  describe('getCCINistMappings', () => {
    it('should return NIST mappings for valid CCI ID', () => {
      const mappings = getCCINistMappings('CCI-000001');
      expect(mappings).toBeDefined();
      expect(Array.isArray(mappings)).toBe(true);
      expect(mappings!.length).toBeGreaterThan(0);
      expect(mappings).toContain('AC-1 a');
    });

    it('should return undefined for invalid CCI ID', () => {
      const mappings = getCCINistMappings('CCI-999999');
      expect(mappings).toBeUndefined();
    });

    it('should return all NIST mappings for CCI with multiple mappings', () => {
      const mappings = getCCINistMappings('CCI-000001');
      expect(mappings).toBeDefined();
      expect(mappings!.length).toBeGreaterThanOrEqual(1);
    });

    it('should handle empty string', () => {
      const mappings = getCCINistMappings('');
      expect(mappings).toBeUndefined();
    });
  });

  describe('getAllCCIIds', () => {
    it('should return array of all CCI IDs', () => {
      const ids = getAllCCIIds();
      expect(Array.isArray(ids)).toBe(true);
      expect(ids.length).toBeGreaterThan(3000); // We know there are 3551 CCIs
    });

    it('should return CCI IDs in expected format', () => {
      const ids = getAllCCIIds();
      const sampleId = ids[0];
      expect(sampleId).toMatch(/^CCI-\d{6}$/);
    });

    it('should return unique CCI IDs', () => {
      const ids = getAllCCIIds();
      const uniqueIds = new Set(ids);
      expect(uniqueIds.size).toBe(ids.length);
    });
  });

  describe('cciExists', () => {
    it('should return true for valid CCI ID', () => {
      expect(cciExists('CCI-000001')).toBe(true);
    });

    it('should return false for invalid CCI ID', () => {
      expect(cciExists('CCI-999999')).toBe(false);
    });

    it('should return false for empty string', () => {
      expect(cciExists('')).toBe(false);
    });

    it('should be case-sensitive', () => {
      expect(cciExists('cci-000001')).toBe(false);
    });
  });

  describe('type guard: non-string inputs', () => {
    it('getCCIDescription returns undefined for null', () => {
      expect(getCCIDescription(null as unknown as string)).toBeUndefined();
    });

    it('getCCIDescription returns undefined for undefined', () => {
      expect(getCCIDescription(undefined as unknown as string)).toBeUndefined();
    });

    it('getCCIDescription returns undefined for number', () => {
      expect(getCCIDescription(123 as unknown as string)).toBeUndefined();
    });

    it('getCCIDescription returns undefined for false', () => {
      expect(getCCIDescription(false as unknown as string)).toBeUndefined();
    });

    it('getCCINistMappings returns undefined for null', () => {
      expect(getCCINistMappings(null as unknown as string)).toBeUndefined();
    });

    it('getCCINistMappings returns undefined for 0', () => {
      expect(getCCINistMappings(0 as unknown as string)).toBeUndefined();
    });

    it('cciExists returns false for null', () => {
      expect(cciExists(null as unknown as string)).toBe(false);
    });

    it('cciExists returns false for 42', () => {
      expect(cciExists(42 as unknown as string)).toBe(false);
    });
  });

  describe('lazy initialization (cold start)', () => {
    it('loads data on first call after module reset', async () => {
      vi.resetModules();
      const { getCCIDescription: getCCIDescFresh } = await import('../src/cci/index.js');
      expect(getCCIDescFresh('CCI-000001')).toBeDefined();
    });
  });

  describe('getNistCCIMappings', () => {
    it('should return CCI IDs for a known NIST control', () => {
      const ccis = getNistCCIMappings('AC-3');
      expect(ccis).toBeDefined();
      expect(Array.isArray(ccis)).toBe(true);
      expect(ccis!.length).toBeGreaterThan(0);
      expect(ccis![0]).toMatch(/^CCI-\d{6}$/);
    });

    it('should normalize qualified controls to base control', () => {
      // 'AC-6 (2)' should be looked up as 'AC-6'
      const fromQualified = getNistCCIMappings('AC-6 (2)');
      const fromBase = getNistCCIMappings('AC-6');
      expect(fromQualified).toEqual(fromBase);
    });

    it('should normalize space-delimited enhancements to base control', () => {
      // 'AC-4 a 1' should be looked up as 'AC-4'
      const fromEnhancement = getNistCCIMappings('AC-4 a 1');
      const fromBase = getNistCCIMappings('AC-4');
      expect(fromEnhancement).toEqual(fromBase);
    });

    it('should return undefined for unknown NIST control', () => {
      expect(getNistCCIMappings('ZZ-999')).toBeUndefined();
    });

    it('should return undefined for empty string', () => {
      expect(getNistCCIMappings('')).toBeUndefined();
    });

    it('should return undefined for null', () => {
      expect(getNistCCIMappings(null as unknown as string)).toBeUndefined();
    });

    it('should return undefined for non-string input', () => {
      expect(getNistCCIMappings(42 as unknown as string)).toBeUndefined();
    });
  });

  describe('nistToCci', () => {
    it('should return CCI IDs for an array of NIST controls', () => {
      const ccis = nistToCci(['AC-3', 'AU-12']);
      expect(Array.isArray(ccis)).toBe(true);
      expect(ccis.length).toBeGreaterThan(0);
      expect(ccis[0]).toMatch(/^CCI-\d{6}$/);
    });

    it('should return sorted results', () => {
      const ccis = nistToCci(['AC-3', 'AU-12']);
      const sorted = [...ccis].sort();
      expect(ccis).toEqual(sorted);
    });

    it('should deduplicate CCI IDs', () => {
      // Same control twice should not produce duplicates
      const ccis = nistToCci(['AC-3', 'AC-3']);
      const unique = new Set(ccis);
      expect(unique.size).toBe(ccis.length);
    });

    it('should return empty array for unknown controls', () => {
      const ccis = nistToCci(['ZZ-999', 'ZZ-888']);
      expect(ccis).toEqual([]);
    });

    it('should return empty array for empty input', () => {
      const ccis = nistToCci([]);
      expect(ccis).toEqual([]);
    });

    it('should skip controls with no mappings', () => {
      const ccisWithUnknown = nistToCci(['AC-3', 'ZZ-999']);
      const ccisWithoutUnknown = nistToCci(['AC-3']);
      expect(ccisWithUnknown).toEqual(ccisWithoutUnknown);
    });
  });

  describe('Edge cases and error handling', () => {
    it('should handle CCI IDs with different number formats', () => {
      // Test some known CCI IDs
      expect(cciExists('CCI-000001')).toBe(true);
      expect(cciExists('CCI-001545')).toBe(true);
      expect(cciExists('CCI-001546')).toBe(true);
    });

    it('should return consistent results for same CCI ID', () => {
      const desc1 = getCCIDescription('CCI-000001');
      const desc2 = getCCIDescription('CCI-000001');
      expect(desc1).toBe(desc2);

      const mappings1 = getCCINistMappings('CCI-000001');
      const mappings2 = getCCINistMappings('CCI-000001');
      expect(mappings1).toEqual(mappings2);
    });
  });
});

File: hdf-mappings/go/cci/cci_test.go
package cci

import (
	"testing"
)

func TestGetCCIDescription(t *testing.T) {
	tests := []struct {
		name     string
		cciID    string
		expected string
		isEmpty  bool
	}{
		{
			name:     "Valid CCI ID",
			cciID:    "CCI-000001",
			expected: "The organization develops an access control policy",
			isEmpty:  false,
		},
		{
			name:     "Another valid CCI ID",
			cciID:    "CCI-000366",
			expected: "",
			isEmpty:  false,
		},
		{
			name:     "Empty CCI ID",
			cciID:    "",
			expected: "",
			isEmpty:  true,
		},
		{
			name:     "Non-existent CCI ID",
			cciID:    "CCI-999999",
			expected: "",
			isEmpty:  true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := GetCCIDescription(tt.cciID)

			if tt.isEmpty {
				if result != "" {
					t.Errorf("Expected empty string, got: %s", result)
				}
			} else if tt.expected != "" && len(result) > 0 {
				// Check if result contains expected substring for valid CCIs
				if len(result) < len(tt.expected) {
					t.Errorf("Description too short. Expected to contain '%s', got: %s", tt.expected, result)
				}
			}
		})
	}
}

func TestGetCCINistMappings(t *testing.T) {
	tests := []struct {
		name       string
		cciID      string
		expectNil  bool
		expectLen  int
		expectContains string
	}{
		{
			name:           "CCI-000366 maps to CM-6 controls",
			cciID:          "CCI-000366",
			expectNil:      false,
			expectLen:      3,
			expectContains: "CM-6 b",
		},
		{
			name:          "CCI-000001 maps to AC-1 controls",
			cciID:         "CCI-000001",
			expectNil:     false,
			expectLen:     3,
			expectContains: "AC-1 a",
		},
		{
			name:      "Empty CCI ID returns nil",
			cciID:     "",
			expectNil: true,
		},
		{
			name:      "Non-existent CCI ID returns nil",
			cciID:     "CCI-999999",
			expectNil: true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := GetCCINistMappings(tt.cciID)

			if tt.expectNil {
				if result != nil {
					t.Errorf("Expected nil, got: %v", result)
				}
				return
			}

			if result == nil {
				t.Errorf("Expected non-nil result for %s", tt.cciID)
				return
			}

			if len(result) != tt.expectLen {
				t.Errorf("Expected %d mappings for %s, got %d: %v", tt.expectLen, tt.cciID, len(result), result)
			}

			if tt.expectContains != "" {
				found := false
				for _, mapping := range result {
					if mapping == tt.expectContains {
						found = true
						break
					}
				}
				if !found {
					t.Errorf("Expected result to contain '%s', got: %v", tt.expectContains, result)
				}
			}
		})
	}
}

func TestGetNistCCIMappings(t *testing.T) {
	tests := []struct {
		name        string
		nistControl string
		expectNil   bool
		expect      []string
	}{
		{
			name:        "SI-10 returns curated CCI",
			nistControl: "SI-10",
			expect:      []string{"CCI-001310"},
		},
		{
			name:        "AC-3 returns curated CCI",
			nistControl: "AC-3",
			expect:      []string{"CCI-000213"},
		},
		{
			name:        "SI-11 returns curated CCI",
			nistControl: "SI-11",
			expect:      []string{"CCI-001312"},
		},
		{
			name:        "SA-11 returns curated CCI",
			nistControl: "SA-11",
			expect:      []string{"CCI-003173"},
		},
		{
			name:        "AC-4 returns multiple curated CCIs",
			nistControl: "AC-4",
			expect:      []string{"CCI-001368", "CCI-001414"},
		},
		{
			name:        "Qualified control normalizes to base",
			nistControl: "SI-10 a 1",
			expect:      []string{"CCI-001310"},
		},
		{
			name:        "Control with enhancement normalizes to base",
			nistControl: "AC-3 (2)",
			expect:      []string{"CCI-000213"},
		},
		{
			name:        "Empty control returns nil",
			nistControl: "",
			expectNil:   true,
		},
		{
			name:        "Control not in curated table returns nil",
			nistControl: "ZZ-999",
			expectNil:   true,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := GetNistCCIMappings(tt.nistControl)

			if tt.expectNil {
				if result != nil {
					t.Errorf("Expected nil, got: %v", result)
				}
				return
			}

			if result == nil {
				t.Errorf("Expected %v for %s, got nil", tt.expect, tt.nistControl)
				return
			}

			if len(result) != len(tt.expect) {
				t.Errorf("Expected %v for %s, got %v", tt.expect, tt.nistControl, result)
				return
			}

			for i, cci := range tt.expect {
				if result[i] != cci {
					t.Errorf("Expected result[%d]=%s for %s, got %s", i, cci, tt.nistControl, result[i])
				}
			}
		})
	}
}

func TestCCIExists(t *testing.T) {
	tests := []struct {
		name     string
		cciID    string
		expected bool
	}{
		{
			name:     "Existing CCI ID",
			cciID:    "CCI-000001",
			expected: true,
		},
		{
			name:     "Another existing CCI ID",
			cciID:    "CCI-000366",
			expected: true,
		},
		{
			name:     "Non-existent CCI ID",
			cciID:    "CCI-999999",
			expected: false,
		},
		{
			name:     "Empty CCI ID",
			cciID:    "",
			expected: false,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := CCIExists(tt.cciID)
			if result != tt.expected {
				t.Errorf("Expected %v for %s, got %v", tt.expected, tt.cciID, result)
			}
		})
	}
}

func TestGetAllCCIIDs(t *testing.T) {
	ids := GetAllCCIIDs()

	if len(ids) == 0 {
		t.Error("Expected non-empty list of CCI IDs")
	}

	// Check that we have a reasonable number of CCIs (should be thousands)
	if len(ids) < 1000 {
		t.Errorf("Expected at least 1000 CCI IDs, got %d", len(ids))
	}

	// Verify some known CCIs are present
	knownCCIs := []string{"CCI-000001", "CCI-000366"}
	for _, cci := range knownCCIs {
		found := false
		for _, id := range ids {
			if id == cci {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("Expected to find %s in CCI ID list", cci)
		}
	}
}

func TestCCIToNIST(t *testing.T) {
	t.Run("maps known CCI IDs to NIST controls", func(t *testing.T) {
		result := CCIToNIST([]string{"CCI-000366"})
		if len(result) == 0 {
			t.Fatal("Expected non-empty NIST controls for CCI-000366")
		}
		found := false
		for _, c := range result {
			if c == "CM-6 b" {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("Expected result to contain 'CM-6 b', got: %v", result)
		}
	})

	t.Run("deduplicates controls from multiple CCI IDs", func(t *testing.T) {
		// Both CCI-000001 and itself map to the same controls; dupes must collapse
		single := CCIToNIST([]string{"CCI-000001"})
		double := CCIToNIST([]string{"CCI-000001", "CCI-000001"})
		if len(single) != len(double) {
			t.Errorf("Expected deduplication: single=%d, double=%d", len(single), len(double))
		}
	})

	t.Run("result is sorted", func(t *testing.T) {
		result := CCIToNIST([]string{"CCI-000366", "CCI-000001"})
		for i := 1; i < len(result); i++ {
			if result[i] < result[i-1] {
				t.Errorf("Result not sorted at index %d: %v", i, result)
			}
		}
	})

	t.Run("empty input returns empty slice", func(t *testing.T) {
		result := CCIToNIST([]string{})
		if result == nil {
			t.Error("Expected non-nil empty slice")
		}
		if len(result) != 0 {
			t.Errorf("Expected empty slice, got: %v", result)
		}
	})

	t.Run("unknown CCI IDs return empty slice", func(t *testing.T) {
		result := CCIToNIST([]string{"CCI-999999"})
		if result == nil {
			t.Error("Expected non-nil empty slice")
		}
		if len(result) != 0 {
			t.Errorf("Expected empty slice, got: %v", result)
		}
	})
}

func TestNISTToCCI(t *testing.T) {
	t.Run("maps known NIST control to CCI IDs", func(t *testing.T) {
		result := NISTToCCI([]string{"AC-3"})
		if len(result) == 0 {
			t.Fatal("Expected non-empty CCI IDs for AC-3")
		}
		found := false
		for _, c := range result {
			if c == "CCI-000213" {
				found = true
				break
			}
		}
		if !found {
			t.Errorf("Expected result to contain 'CCI-000213', got: %v", result)
		}
	})

	t.Run("deduplicates CCI IDs from multiple NIST controls", func(t *testing.T) {
		single := NISTToCCI([]string{"AC-3"})
		double := NISTToCCI([]string{"AC-3", "AC-3"})
		if len(single) != len(double) {
			t.Errorf("Expected deduplication: single=%d, double=%d", len(single), len(double))
		}
	})

	t.Run("result is sorted", func(t *testing.T) {
		result := NISTToCCI([]string{"AC-3", "SI-10"})
		for i := 1; i < len(result); i++ {
			if result[i] < result[i-1] {
				t.Errorf("Result not sorted at index %d: %v", i, result)
			}
		}
	})

	t.Run("empty input returns empty slice", func(t *testing.T) {
		result := NISTToCCI([]string{})
		if result == nil {
			t.Error("Expected non-nil empty slice")
		}
		if len(result) != 0 {
			t.Errorf("Expected empty slice, got: %v", result)
		}
	})

	t.Run("unknown NIST control returns empty slice", func(t *testing.T) {
		result := NISTToCCI([]string{"ZZ-999"})
		if result == nil {
			t.Error("Expected non-nil empty slice")
		}
		if len(result) != 0 {
			t.Errorf("Expected empty slice, got: %v", result)
		}
	})

	t.Run("qualified control normalizes before lookup", func(t *testing.T) {
		base := NISTToCCI([]string{"SI-10"})
		qualified := NISTToCCI([]string{"SI-10 a 1"})
		if len(base) != len(qualified) {
			t.Errorf("Expected same result for base and qualified: %v vs %v", base, qualified)
		}
	})
}

func TestLazyLoading(t *testing.T) {
	// First call should load data
	desc1 := GetCCIDescription("CCI-000001")
	if desc1 == "" {
		t.Error("Expected description for CCI-000001")
	}

	// Second call should use cached data
	desc2 := GetCCIDescription("CCI-000001")
	if desc1 != desc2 {
		t.Error("Expected same description on second call (cached)")
	}
}
